src/layout/umap.c

```diff
@@ -92,27 +150,82 @@ static igraph_error_t igraph_i_umap_find_sigma(const igraph_vector_t *distances,
+ * making it also a useful graph layout algorithm. See \ref igraph_layout_umap() for more information.
+ *
+ * </para><para>
+ *
+ * An early step in UMAP is to compute exponentially decaying "weights" from the
+ * distance graph. Connectivities can also be viewed as edge weights that quantify
+ * similarity between two vertices. This function computes weights from the
+ * distance graph. To compute the layout from precomputed weights, call
+ * igraph_layout_umap with the "distances_are_weights" argument set to true.
+ *
+ * </para><para>
+ *
+ * Technical note: For each vertex, this function computes its scale factor (sigma),
+ * its connectivity correction (rho), and finally the weights themselves.
+ *
+ * </para><para>
+ * References:
+ *
+ * </para><para>
+ * Leland McInnes, John Healy, and James Melville. https://arxiv.org/abs/1802.03426
+ *
+ * \param graph Pointer to the distance graph.
+ * \param distances Pointer to the vector with the vertex-to-vertex distance associated with
+ *   each edge.
+ * \param weights Pointer to an initialized vector where the result will be stored.
+ *
+ * \return Error code.
+ */
+igraph_error_t igraph_layout_umap_compute_weights(
+        const igraph_t *graph,
+        const igraph_vector_t *distances,
+        igraph_vector_t *weights) {
 
     igraph_integer_t no_of_nodes = igraph_vcount(graph);
     igraph_integer_t no_of_edges = igraph_ecount(graph);
     igraph_integer_t no_of_neis, eid;
     igraph_vector_int_t eids, weight_seen;
     igraph_real_t rho, dist_max, dist, sigma, weight, weight_inv, sigma_target;
 
+    /* UMAP is sometimes used on unweighted graphs. Otherwise, check distance vector */
+    if (distances != NULL) {
+        if (igraph_vector_size(distances) != no_of_edges) {
+            IGRAPH_ERROR("Distances must be the same number as the edges in the graph.", IGRAPH_EINVAL);
+        }
+
+        if (no_of_edges > 0) {
+            igraph_real_t distance_min = igraph_vector_min(distances);
+            if (distance_min < 0) {
+                IGRAPH_ERROR("Distance vector must be nonnegative.", IGRAPH_EINVAL);
+            } else if (isnan(distance_min)) {
+                IGRAPH_ERROR("Distance vector must not contain NaN values.", IGRAPH_EINVAL);
+            }
+        }
+    }
+
+    /* reserve memory for the weights */
+    IGRAPH_CHECK(igraph_vector_resize(weights, no_of_edges));
+
     /* if the original graph is unweighted, probabilities are 1 throughout */
     if (distances == NULL) {
         for (igraph_integer_t j = 0; j < no_of_edges; j++) {
-            VECTOR(*umap_weights)[j] = 1;
+            VECTOR(*weights)[j] = 1;
         }
         return IGRAPH_SUCCESS;
     }
-    /* alright, the graph is weighted */
+    /* alright, the graph has actual distances */
 
-    /* Initialize vectors and matrices */
+    /* Initialize auxiliary vectors */
     IGRAPH_VECTOR_INT_INIT_FINALLY(&eids, 0);
     IGRAPH_VECTOR_INT_INIT_FINALLY(&weight_seen, no_of_edges);
```

Comment:
Minor question: This is an integer vector, but it is used as a Boolean in practice, i.e. the counts are recorded but it's only ever checked whether they're zero or not.  Are you planning to use the counts later, or can it be changed to an actual Boolean vector?

Reply:
I should make it a bool